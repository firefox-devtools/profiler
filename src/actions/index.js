/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import * as app from './app';
import * as icons from './icons';
import * as profileView from './profile-view';
import * as publish from './publish';
import * as receiveProfile from './receive-profile';
import * as zippedProfiles from './zipped-profiles';
import * as code from './code';

export default Object.assign(
  {},
  app,
  icons,
  profileView,
  publish,
  receiveProfile,
  zippedProfiles,
  code
);
