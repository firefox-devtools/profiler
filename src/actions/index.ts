/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
import * as app from './app';
import * as icons from './icons';
import * as l10n from './l10n';
import * as profileView from './profile-view';
import * as publish from './publish';
import * as receiveProfile from './receive-profile';
import * as zippedProfiles from './zipped-profiles';
import * as code from './code';

export default Object.assign(
  {},
  app,
  icons,
  l10n,
  profileView,
  publish,
  receiveProfile,
  zippedProfiles,
  code
);
