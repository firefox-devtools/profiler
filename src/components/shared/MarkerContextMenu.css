/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

/* Most of the CSS logic to be able to add icons is part of the more generic
 * file ContextMenu.css that is applied for all context menus. Especially the
 * CSS class `react-contextmenu-icon` is used in addition to these more specific
 * classes, and that's where other properties for backgrounds are specified.
 */

/* Here are all icons for the marker context menu. */
.markerContextMenuIconSetSelectionFromMarker {
  background-image: url(../../../res/img/svg/set-selection-from-marker.svg);
}

.markerContextMenuIconStartSelectionHere, /* <- for dot markers */
.markerContextMenuIconStartSelectionAtMarkerStart /* <- for interval markers */ {
  background-image: url(../../../res/img/svg/start-selection-at-marker-start.svg);
}

.markerContextMenuIconStartSelectionAtMarkerEnd {
  background-image: url(../../../res/img/svg/start-selection-at-marker-end.svg);
}

.markerContextMenuIconEndSelectionAtMarkerStart {
  background-image: url(../../../res/img/svg/end-selection-at-marker-start.svg);
}

.markerContextMenuIconEndSelectionHere, /* <- for dot markers */
.markerContextMenuIconEndSelectionAtMarkerEnd /* <- for interval markers */ {
  background-image: url(../../../res/img/svg/end-selection-at-marker-end.svg);
}

.markerContextMenuIconCopyDescription {
  background-image: url(../../../res/img/svg/copy-dark.svg);
}

.markerContextMenuIconCopyStack {
  background-image: url(../../../res/img/svg/copy-marker-stack.svg);
}

.markerContextMenuIconCopyUrl {
  background-image: url(../../../res/img/svg/globe-dark.svg);
}

.markerContextMenuIconCopyPayload {
  background-image: url(../../../res/img/svg/copy-marker-payload.svg);
}

.markerContextMenuSelectThread {
  background-image: url(../../../res/img/svg/select-thread.svg);
}

/* These icons don't look nice when selected, unless if we invert them. However
 * the other icons look better without the filter, so we don't change them in
 * this state. */
.react-contextmenu-item--selected .markerContextMenuIconCopyDescription,
.react-contextmenu-item:hover .markerContextMenuIconCopyDescription,
.react-contextmenu-item--selected .markerContextMenuIconCopyUrl,
.react-contextmenu-item:hover .markerContextMenuIconCopyUrl,
.react-contextmenu-item--selected .markerContextMenuIconCopyStack,
.react-contextmenu-item:hover .markerContextMenuIconCopyStack,
.react-contextmenu-item--selected .markerContextMenuIconCopyPayload,
.react-contextmenu-item:hover .markerContextMenuIconCopyPayload,
.react-contextmenu-item--selected .markerContextMenuSelectThread,
.react-contextmenu-item:hover .markerContextMenuSelectThread {
  filter: invert(1);
}
