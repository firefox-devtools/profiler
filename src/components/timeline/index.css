/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

.timelineOverflowEdgeIndicatorScrollbox {
  position: absolute;

  /* if the platform scrollbar ends up being larger than --vertical-scrollbar-reserved-width (which is unexpected), make sure there is no horizontal scrollbar */
  overflow: hidden auto;
  width: 100%;
  height: 100%;
  padding-left: var(--thread-label-column-width);
  margin-right: calc(var(--vertical-scrollbar-reserved-width) * -1);
  margin-left: calc(var(--thread-label-column-width) * -1);
  background: linear-gradient(
    to left,
    white,
    #eee calc(var(--vertical-scrollbar-reserved-width) - 1px),
    var(--grey-30) 0,
    var(--grey-30) var(--vertical-scrollbar-reserved-width),
    transparent 0
  );
}

.timelineThreadList {
  width: calc(100vw - var(--vertical-scrollbar-reserved-width));
  padding: 0;
  margin: 0 0 0 calc(var(--thread-label-column-width) * -1);
  background-color: var(--grey-20);
  box-shadow: inset 0 1px var(--grey-30);
  list-style: none;
}

.timelineSettingsHiddenTracks {
  /* Make sure and override all rules that have to do with button defaults. */
  position: relative;
  width: var(--thread-label-column-width);

  /* 8px is intentional to align this text with the text above. */
  padding: 0 20px 0 8px;
  border: 0;

  /* Separator line between tracks menu button and timeline ruler */
  border-right: 1px solid var(--grey-30);
  background: none;
  color: inherit;
  font: inherit;
  text-align: left;
}

.timelineSettingsHiddenTracks:hover {
  background-color: rgb(0 0 0 / 0.1);
}

.timelineSettingsHiddenTracks:hover:active {
  background-color: rgb(0 0 0 / 0.2);
}

/* This is the dropdown arrow on the left of the button. */
.timelineSettingsHiddenTracks::before {
  position: absolute;
  top: 2px;
  right: 2px;
  border-top: 6px solid;
  border-right: 4px solid transparent;
  border-bottom: 0 solid transparent;
  border-left: 4px solid transparent;
  margin-top: 5px;
  margin-right: 5px;
  margin-left: 5px;
  color: var(--grey-60);
  content: '';

  /* Opt-out of forced colors so arrow is visible in High Contrast Mode */
  forced-color-adjust: none;
}

.timelineSettingsHiddenTracksNumber {
  font-weight: bold;
}

.timelineHeader {
  /* This should be on top of the overflow edge indicator that wraps the thread
   list, so border bottom will always be visible. */
  z-index: 1;
  display: flex;
  height: 20px;
  flex-direction: row;

  /* Timeline header content should end before the scrollbar but also border
   * bottom should extend to the whole timeline header. Because of that, we are
   * creating a padding and a minus margin with the scrollbar width to make
   * this work. */
  padding-right: var(--vertical-scrollbar-reserved-width);
  border-bottom: 1px solid var(--grey-30);
  margin-right: calc(var(--vertical-scrollbar-reserved-width) * -1);
  margin-left: calc(var(--thread-label-column-width) * -1);

  /* This div needs to be apaque so that it can properly hide the hidden tracks
   * container border below */
  background: white;
}

.tracksContainer {
  /* This should be below the timeline header. */
  z-index: 0;

  /* Moving the timeline 1 pixel below the timeline header. This way, the first
   * visible track's top border will be hidden as expected. */
  margin-top: -1px;
  margin-bottom: -1px;
}

@media (forced-colors: active) {
  .timelineSettingsHiddenTracks {
    border: 1px solid currentcolor;
    background-color: ButtonFace;
    color: ButtonText;
  }

  .timelineSettingsHiddenTracks:hover {
    background-color: SelectedItemText;
    color: SelectedItem;
  }

  .timelineSettingsHiddenTracks:hover:active {
    border-color: ButtonText;
  }

  .timelineSettingsHiddenTracks::before {
    color: currentcolor;
  }
}
