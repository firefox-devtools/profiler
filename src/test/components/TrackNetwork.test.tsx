/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import { Provider } from 'react-redux';

// This module is mocked.
import copy from 'copy-to-clipboard';

import {
  render,
  fireEvent,
  screen,
  act,
} from 'firefox-profiler/test/fixtures/testing-library';

import { TrackNetwork } from 'firefox-profiler/components/timeline/TrackNetwork';
import { MaybeMarkerContextMenu } from 'firefox-profiler/components/shared/MarkerContextMenu';
import {
  TRACK_NETWORK_ROW_HEIGHT,
  TRACK_NETWORK_ROW_REPEAT,
} from 'firefox-profiler/app-logic/constants';
import { ensureExists } from 'firefox-profiler/utils/types';
import { selectedThreadSelectors } from 'firefox-profiler/selectors/per-thread';
import { changeSelectedNetworkMarker } from 'firefox-profiler/actions/profile-view';

import { autoMockCanvasContext } from '../fixtures/mocks/canvas-context';
import { triggerResizeObservers } from '../fixtures/mocks/resize-observer';
import { mockRaf } from '../fixtures/mocks/request-animation-frame';
import { storeWithProfile } from '../fixtures/stores';
import {
  getMouseEvent,
  fireFullClick,
  fireFullContextMenu,
  addRootOverlayElement,
  removeRootOverlayElement,
} from '../fixtures/utils';
import { getNetworkTrackProfile } from '../fixtures/profiles/processed-profile';
import {
  autoMockElementSize,
  setMockedElementSize,
} from '../fixtures/mocks/element-size';

// The graph is 400 pixels wide based on the element size mock, and the graph
// height mimicks what is computed by the actual component.
const GRAPH_WIDTH = 400;
const GRAPH_HEIGHT = TRACK_NETWORK_ROW_HEIGHT * TRACK_NETWORK_ROW_REPEAT;

autoMockCanvasContext();
autoMockElementSize({ width: GRAPH_WIDTH, height: GRAPH_HEIGHT });

beforeEach(addRootOverlayElement);
afterEach(removeRootOverlayElement);

describe('timeline/TrackNetwork', function () {
  it('matches the component snapshot', () => {
    const { container, unmount } = setup();
    expect(container.firstChild).toMatchSnapshot();
    // Trigger any unmounting behavior handlers, just make sure it doesn't
    // throw any errors.
    unmount();
  });

  it('matches the 2d context snapshot', () => {
    const { getContextDrawCalls } = setup();
    expect(getContextDrawCalls()).toMatchSnapshot();
  });

  it('redraws on a resize', () => {
    const { getContextDrawCalls } = setup();
    // Flush out any existing draw calls.
    getContextDrawCalls();
    // Ensure we start out with 0.
    expect(getContextDrawCalls()).toHaveLength(0);

    // Send out the resize with a width change.
    // By changing the "fake" result of getBoundingClientRect, we ensure that
    // the pure components rerender because their `width` props change.
    setMockedElementSize({ width: GRAPH_WIDTH - 100, height: GRAPH_HEIGHT });
    triggerResizeObservers();
    expect(getContextDrawCalls()).not.toHaveLength(0);
  });

  it('draws differently a request and displays a tooltip when hovered', () => {
    const { getContextDrawCalls } = setup();
    // Flush out any existing draw calls.
    getContextDrawCalls();

    const canvas = ensureExists(document.querySelector('canvas'));
    fireEvent(
      canvas,
      getMouseEvent('mousemove', {
        pageX: 12,
        offsetX: 12,
        pageY: 2,
        offsetY: 2,
      })
    );

    const tooltip = screen.getByTestId('tooltip');
    expect(tooltip).toMatchSnapshot();

    const drawCalls = getContextDrawCalls();
    // We draw at least one hovered request with a different stroke style.
    expect(drawCalls).toContainEqual(['set strokeStyle', '#0069aa']);
    expect(drawCalls).toMatchSnapshot();
  });

  it('changes the hovered marker information in the redux store', () => {
    const { getState } = setup();

    const canvas = ensureExists(document.querySelector('canvas'));

    // Hover marker 0
    fireEvent(
      canvas,
      getMouseEvent('mousemove', {
        pageX: 12,
        offsetX: 12,
        pageY: 2,
        offsetY: 2,
      })
    );

    expect(selectedThreadSelectors.getHoveredMarkerIndex(getState())).toBe(0);

    // Hover marker 1
    fireEvent(
      canvas,
      getMouseEvent('mousemove', {
        pageX: 30,
        offsetX: 30,
        pageY: 7,
        offsetY: 7,
      })
    );
    expect(selectedThreadSelectors.getHoveredMarkerIndex(getState())).toBe(1);

    // Hover some blank space
    fireEvent(
      canvas,
      getMouseEvent('mousemove', {
        pageX: 12,
        offsetX: 12,
        pageY: 10,
        offsetY: 10,
      })
    );
    expect(selectedThreadSelectors.getHoveredMarkerIndex(getState())).toBe(
      null
    );
  });

  it('displays a context menu when right clicking', () => {
    // Always use fake timers when dealing with context menus.
    jest.useFakeTimers();

    const { getContextMenu, clickOnMenuItem, getContextDrawCalls } = setup();
    const canvas = ensureExists(document.querySelector('canvas'));

    // First the user hovers the track.
    fireEvent(
      canvas,
      getMouseEvent('mousemove', {
        pageX: 12,
        offsetX: 12,
        pageY: 2,
        offsetY: 2,
      })
    );

    expect(screen.getByTestId('tooltip')).toBeInTheDocument();

    // Flush existing draw calls, to check that we redraw properly.
    getContextDrawCalls();

    // Then the user right clicks.
    fireFullContextMenu(canvas);

    expect(getContextMenu()).toBeInTheDocument();
    expect(screen.queryByTestId('tooltip')).not.toBeInTheDocument();

    // Check that we redrew with a different style.
    const drawCalls = getContextDrawCalls();
    // We draw at least one hovered request with a different stroke style.
    expect(drawCalls).toContainEqual(['set strokeStyle', '#0069aa']);

    // The user clicks on an item.
    clickOnMenuItem('Copy description');
    expect(copy).toHaveBeenLastCalledWith('Load 0: https://mozilla.org');
    expect(getContextMenu()).not.toHaveClass('react-contextmenu--visible');

    act(() => jest.runAllTimers());

    expect(document.querySelector('.react-contextmenu')).toBeFalsy();
  });

  it('displays a context menu when right clicking, and hides it when clicking in blank space', () => {
    // Always use fake timers when dealing with context menus.
    jest.useFakeTimers();

    const { getContextMenu } = setup();
    const canvas = ensureExists(document.querySelector('canvas'));

    // First the user hovers the track.
    fireEvent(
      canvas,
      getMouseEvent('mousemove', {
        pageX: 12,
        offsetX: 12,
        pageY: 2,
        offsetY: 2,
      })
    );

    expect(screen.getByTestId('tooltip')).toBeInTheDocument();

    // Then the user right clicks.
    fireFullContextMenu(canvas);

    expect(getContextMenu()).toBeInTheDocument();
    expect(screen.queryByTestId('tooltip')).not.toBeInTheDocument();

    // The user hovers some blank space
    fireEvent(
      canvas,
      getMouseEvent('mousemove', {
        pageX: 12,
        offsetX: 12,
        pageY: 20,
        offsetY: 20,
      })
    );

    // And then right clicks again => In this case the context menu should disappear!
    fireFullContextMenu(canvas);
    act(() => jest.runAllTimers());
    expect(document.querySelector('.react-contextmenu')).toBeFalsy();
  });

  it('selects the network marker when left clicking', () => {
    const { getContextDrawCalls, getState } = setup();

    const canvas = ensureExists(document.querySelector('canvas'));

    // First the user hovers the track.
    fireEvent(
      canvas,
      getMouseEvent('mousemove', {
        pageX: 12,
        pageY: 2,
      })
    );

    // Check that nothing is selected yet.
    expect(
      selectedThreadSelectors.getSelectedNetworkMarkerIndex(getState())
    ).toBe(null);

    // Then the user left clicks.
    fireFullClick(canvas);
    expect(
      selectedThreadSelectors.getSelectedNetworkMarkerIndex(getState())
    ).toBe(0);

    // Flush out any existing draw calls as we're interested in what comes news.
    getContextDrawCalls();
    // Ensure we start out with 0.
    expect(getContextDrawCalls().length).toEqual(0);
    getContextDrawCalls();

    // The mouse leaves the track.
    fireEvent.mouseLeave(canvas);

    const drawCalls = getContextDrawCalls();
    // We draw at least one hovered request with a different stroke style.
    expect(drawCalls).toContainEqual(['set strokeStyle', '#0069aa']);
  });

  it('draws the selected network marker when it changes elsewhere', () => {
    const { getContextDrawCalls, getState, dispatch } = setup();

    // Flush out any existing draw calls.
    getContextDrawCalls();
    // Ensure we start out with 0.
    expect(getContextDrawCalls().length).toEqual(0);

    // Check that nothing is selected yet.
    expect(
      selectedThreadSelectors.getSelectedNetworkMarkerIndex(getState())
    ).toBe(null);

    act(() => {
      dispatch(changeSelectedNetworkMarker(0, 2));
    });

    // Check that we redrew with a selected style.
    const drawCalls = getContextDrawCalls();
    // We draw at least one hovered request with a different stroke style.
    expect(drawCalls).toContainEqual(['set strokeStyle', '#0069aa']);
  });
});

describe('VerticalIndicators', function () {
  it('creates the vertical indicators', function () {
    const { getIndicatorLines, getState } = setup();
    const markerIndexes =
      selectedThreadSelectors.getTimelineVerticalMarkerIndexes(getState());
    const markerCount = 5;
    expect(markerIndexes).toHaveLength(markerCount);
    expect(getIndicatorLines()).toHaveLength(markerCount);
  });

  it('displays tooltips', function () {
    const { getIndicatorLines } = setup();
    const [firstIndicator] = getIndicatorLines();
    fireEvent.mouseOver(firstIndicator);
    fireEvent(
      firstIndicator,
      getMouseEvent('mousemove', {
        pageX: 11,
        pageY: 22,
      })
    );

    const tooltip = screen.getByTestId('tooltip');
    expect(tooltip).toMatchSnapshot();
  });

  it('displays a context menu when right clicking', () => {
    // Always use fake timers when dealing with context menus.
    jest.useFakeTimers();

    const { getIndicatorLines, getContextMenu, clickOnMenuItem } = setup();

    // We're looking at the third indicator only to change from the previous test.
    const [, , thirdIndicator] = getIndicatorLines();
    fireEvent.mouseOver(thirdIndicator);
    // Note: we don't need to specify pageX/pageY here, as we don't test the
    // position of the tooltip, but only its presence.
    fireEvent.mouseMove(thirdIndicator);

    const tooltip = screen.getByTestId('tooltip');
    expect(tooltip).toBeInTheDocument();

    // Then the user right clicks.
    fireFullContextMenu(thirdIndicator);
    expect(getContextMenu()).toBeInTheDocument();
    expect(tooltip).not.toBeInTheDocument();

    // The user clicks on an item.
    clickOnMenuItem('Copy description');
    expect(copy).toHaveBeenLastCalledWith('DOMContentLoaded');
    expect(getContextMenu()).not.toHaveClass('react-contextmenu--visible');

    act(() => jest.runAllTimers());

    expect(document.querySelector('.react-contextmenu')).toBeFalsy();
  });
});

function setup() {
  const profile = getNetworkTrackProfile();
  const store = storeWithProfile(profile);

  const { getState, dispatch } = store;
  const flushRafCalls = mockRaf();

  const renderResult = render(
    <Provider store={store}>
      <TrackNetwork threadIndex={0} />
      <MaybeMarkerContextMenu />
    </Provider>
  );

  const getIndicatorLines = () =>
    screen.getAllByTestId('vertical-indicator-line');

  // WithSize uses requestAnimationFrame
  flushRafCalls();

  /**
   * Coordinate the flushing of the requestAnimationFrame and the draw calls.
   */
  function getContextDrawCalls() {
    flushRafCalls();
    return (window as any).__flushDrawLog();
  }

  function getContextMenu() {
    return ensureExists(
      document.querySelector('.react-contextmenu'),
      `Couldn't find the context menu.`
    );
  }

  function clickOnMenuItem(stringOrRegexp: string | RegExp) {
    fireFullClick(screen.getByText(stringOrRegexp));
  }

  return {
    ...renderResult,
    dispatch,
    getState,
    thread: profile.threads[0],
    store,
    getContextDrawCalls,
    getIndicatorLines,
    getContextMenu,
    clickOnMenuItem,
  };
}
