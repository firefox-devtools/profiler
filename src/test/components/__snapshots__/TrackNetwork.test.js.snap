// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VerticalIndicators displays tooltips 1`] = `
<div
  class="tooltip"
  data-testid="tooltip"
  style="left: 22px; top: 33px;"
>
  <div>
    <span
      class="timelineVerticalIndicatorsSwatch"
    />
     
    Load
    <span
      class="timelineVerticalIndicatorsDim"
    >
       at 
    </span>
    <span
      class="timelineVerticalIndicatorsTime"
    >
      0.003s
    </span>
     
  </div>
  <div
    class="timelineVerticalIndicatorsUrl"
  >
    https://developer.mozilla.org/en-US/
  </div>
</div>
`;

exports[`timeline/TrackNetwork draws differently a request and displays a tooltip when hovered 1`] = `
<div
  class="tooltip"
  data-testid="tooltip"
  style="left: 23px; top: 13px;"
>
  <div
    class="tooltipMarker tooltipNetwork"
  >
    <div
      class="tooltipHeader"
    >
      <div
        class="tooltipOneLine"
      >
        <div
          class="tooltipTiming"
        >
          1ms
           @ 
          0s
        </div>
        <div
          class="tooltipTitle"
        >
          Load 0: https://mozilla.org
        </div>
      </div>
    </div>
    <div
      class="tooltipDetails"
    >
      <div
        class="tooltipLabel"
      >
        Status
        :
      </div>
      Response received
      <div
        class="tooltipLabel"
      >
        URL
        :
      </div>
      <span
        class="tooltipDetailsUrl"
      >
        https://mozilla.org
      </span>
      <div
        class="tooltipLabel"
      >
        Priority
        :
      </div>
      Normal(0)
      <div
        class="tooltipLabel"
      >
        MIME type
        :
      </div>
      <div
        class="tooltipNetworkMimeType"
      >
        <span
          class="tooltipNetworkMimeTypeSwatch colored-square network-color-html"
          title="text/html"
        />
        text/html
      </div>
      <div
        class="tooltipLabel"
      >
        Track
        :
      </div>
      Empty
    </div>
    <div
      class="tooltipNetworkPhases network-color-html"
    >
      <div
        class="tooltipLabel"
      >
        HTTP response
        :
      </div>
      <div
        aria-label="Starting at 0 milliseconds, duration is 1.0 milliseconds"
      >
        1.0ms
      </div>
      <div
        aria-hidden="true"
        class="tooltipNetworkPhase"
        style="margin-left: 0%; margin-right: 0%; opacity: 1;"
      />
    </div>
  </div>
</div>
`;

exports[`timeline/TrackNetwork draws differently a request and displays a tooltip when hovered 2`] = `
Array [
  Array [
    "clearRect",
    0,
    0,
    400,
    35,
  ],
  Array [
    "set strokeStyle",
    "rgba(0, 127, 255, 0.3)",
  ],
  Array [
    "set lineCap",
    "round",
  ],
  Array [
    "set lineWidth",
    3.75,
  ],
  Array [
    "beginPath",
  ],
  Array [
    "moveTo",
    6.666666666666673,
    7.5,
  ],
  Array [
    "lineTo",
    73.33333333333331,
    7.5,
  ],
  Array [
    "stroke",
  ],
  Array [
    "beginPath",
  ],
  Array [
    "moveTo",
    13.333333333333346,
    12.5,
  ],
  Array [
    "lineTo",
    80.00000000000001,
    12.5,
  ],
  Array [
    "stroke",
  ],
  Array [
    "beginPath",
  ],
  Array [
    "moveTo",
    19.99999999999999,
    17.5,
  ],
  Array [
    "lineTo",
    86.66666666666666,
    17.5,
  ],
  Array [
    "stroke",
  ],
  Array [
    "beginPath",
  ],
  Array [
    "moveTo",
    26.666666666666664,
    22.5,
  ],
  Array [
    "lineTo",
    93.33333333333336,
    22.5,
  ],
  Array [
    "stroke",
  ],
  Array [
    "beginPath",
  ],
  Array [
    "moveTo",
    33.333333333333336,
    27.5,
  ],
  Array [
    "lineTo",
    100,
    27.5,
  ],
  Array [
    "stroke",
  ],
  Array [
    "beginPath",
  ],
  Array [
    "moveTo",
    40.00000000000001,
    32.5,
  ],
  Array [
    "lineTo",
    106.66666666666666,
    32.5,
  ],
  Array [
    "stroke",
  ],
  Array [
    "beginPath",
  ],
  Array [
    "moveTo",
    46.66666666666668,
    2.5,
  ],
  Array [
    "lineTo",
    113.33333333333336,
    2.5,
  ],
  Array [
    "stroke",
  ],
  Array [
    "beginPath",
  ],
  Array [
    "moveTo",
    53.33333333333333,
    7.5,
  ],
  Array [
    "lineTo",
    120,
    7.5,
  ],
  Array [
    "stroke",
  ],
  Array [
    "beginPath",
  ],
  Array [
    "moveTo",
    60,
    12.5,
  ],
  Array [
    "lineTo",
    126.6666666666667,
    12.5,
  ],
  Array [
    "stroke",
  ],
  Array [
    "set strokeStyle",
    "#0069aa",
  ],
  Array [
    "beginPath",
  ],
  Array [
    "moveTo",
    0,
    2.5,
  ],
  Array [
    "lineTo",
    66.66666666666667,
    2.5,
  ],
  Array [
    "stroke",
  ],
]
`;

exports[`timeline/TrackNetwork matches the 2d context snapshot 1`] = `
Array [
  Array [
    "clearRect",
    0,
    0,
    0,
    35,
  ],
  Array [
    "set strokeStyle",
    "rgba(0, 127, 255, 0.3)",
  ],
  Array [
    "set lineCap",
    "round",
  ],
  Array [
    "set lineWidth",
    3.75,
  ],
  Array [
    "beginPath",
  ],
  Array [
    "moveTo",
    0,
    2.5,
  ],
  Array [
    "lineTo",
    0,
    2.5,
  ],
  Array [
    "stroke",
  ],
  Array [
    "beginPath",
  ],
  Array [
    "moveTo",
    0,
    7.5,
  ],
  Array [
    "lineTo",
    0,
    7.5,
  ],
  Array [
    "stroke",
  ],
  Array [
    "beginPath",
  ],
  Array [
    "moveTo",
    0,
    12.5,
  ],
  Array [
    "lineTo",
    0,
    12.5,
  ],
  Array [
    "stroke",
  ],
  Array [
    "beginPath",
  ],
  Array [
    "moveTo",
    0,
    17.5,
  ],
  Array [
    "lineTo",
    0,
    17.5,
  ],
  Array [
    "stroke",
  ],
  Array [
    "beginPath",
  ],
  Array [
    "moveTo",
    0,
    22.5,
  ],
  Array [
    "lineTo",
    0,
    22.5,
  ],
  Array [
    "stroke",
  ],
  Array [
    "beginPath",
  ],
  Array [
    "moveTo",
    0,
    27.5,
  ],
  Array [
    "lineTo",
    0,
    27.5,
  ],
  Array [
    "stroke",
  ],
  Array [
    "beginPath",
  ],
  Array [
    "moveTo",
    0,
    32.5,
  ],
  Array [
    "lineTo",
    0,
    32.5,
  ],
  Array [
    "stroke",
  ],
  Array [
    "beginPath",
  ],
  Array [
    "moveTo",
    0,
    2.5,
  ],
  Array [
    "lineTo",
    0,
    2.5,
  ],
  Array [
    "stroke",
  ],
  Array [
    "beginPath",
  ],
  Array [
    "moveTo",
    0,
    7.5,
  ],
  Array [
    "lineTo",
    0,
    7.5,
  ],
  Array [
    "stroke",
  ],
  Array [
    "beginPath",
  ],
  Array [
    "moveTo",
    0,
    12.5,
  ],
  Array [
    "lineTo",
    0,
    12.5,
  ],
  Array [
    "stroke",
  ],
  Array [
    "clearRect",
    0,
    0,
    400,
    35,
  ],
  Array [
    "set strokeStyle",
    "rgba(0, 127, 255, 0.3)",
  ],
  Array [
    "set lineCap",
    "round",
  ],
  Array [
    "set lineWidth",
    3.75,
  ],
  Array [
    "beginPath",
  ],
  Array [
    "moveTo",
    0,
    2.5,
  ],
  Array [
    "lineTo",
    66.66666666666667,
    2.5,
  ],
  Array [
    "stroke",
  ],
  Array [
    "beginPath",
  ],
  Array [
    "moveTo",
    6.666666666666673,
    7.5,
  ],
  Array [
    "lineTo",
    73.33333333333331,
    7.5,
  ],
  Array [
    "stroke",
  ],
  Array [
    "beginPath",
  ],
  Array [
    "moveTo",
    13.333333333333346,
    12.5,
  ],
  Array [
    "lineTo",
    80.00000000000001,
    12.5,
  ],
  Array [
    "stroke",
  ],
  Array [
    "beginPath",
  ],
  Array [
    "moveTo",
    19.99999999999999,
    17.5,
  ],
  Array [
    "lineTo",
    86.66666666666666,
    17.5,
  ],
  Array [
    "stroke",
  ],
  Array [
    "beginPath",
  ],
  Array [
    "moveTo",
    26.666666666666664,
    22.5,
  ],
  Array [
    "lineTo",
    93.33333333333336,
    22.5,
  ],
  Array [
    "stroke",
  ],
  Array [
    "beginPath",
  ],
  Array [
    "moveTo",
    33.333333333333336,
    27.5,
  ],
  Array [
    "lineTo",
    100,
    27.5,
  ],
  Array [
    "stroke",
  ],
  Array [
    "beginPath",
  ],
  Array [
    "moveTo",
    40.00000000000001,
    32.5,
  ],
  Array [
    "lineTo",
    106.66666666666666,
    32.5,
  ],
  Array [
    "stroke",
  ],
  Array [
    "beginPath",
  ],
  Array [
    "moveTo",
    46.66666666666668,
    2.5,
  ],
  Array [
    "lineTo",
    113.33333333333336,
    2.5,
  ],
  Array [
    "stroke",
  ],
  Array [
    "beginPath",
  ],
  Array [
    "moveTo",
    53.33333333333333,
    7.5,
  ],
  Array [
    "lineTo",
    120,
    7.5,
  ],
  Array [
    "stroke",
  ],
  Array [
    "beginPath",
  ],
  Array [
    "moveTo",
    60,
    12.5,
  ],
  Array [
    "lineTo",
    126.6666666666667,
    12.5,
  ],
  Array [
    "stroke",
  ],
]
`;

exports[`timeline/TrackNetwork matches the component snapshot 1`] = `
<div
  class="timelineTrackNetwork"
  style="height: 35px;"
>
  <div
    class="react-contextmenu-wrapper treeViewContextMenu"
  >
    <canvas
      class="timelineTrackNetworkCanvas"
      height="35"
      width="400"
    />
    <div
      class="timelineVerticalIndicatorsLine"
      data-marker-index="10"
      data-testid="vertical-indicator-line"
      style="--vertical-indicator-color: var(--red-60); left: 200px;"
    />
    <div
      class="timelineVerticalIndicatorsLine"
      data-marker-index="11"
      data-testid="vertical-indicator-line"
      style="--vertical-indicator-color: #000; left: 200px;"
    />
    <div
      class="timelineVerticalIndicatorsLine"
      data-marker-index="12"
      data-testid="vertical-indicator-line"
      style="--vertical-indicator-color: var(--blue-50); left: 266.6666666666667px;"
    />
    <div
      class="timelineVerticalIndicatorsLine"
      data-marker-index="13"
      data-testid="vertical-indicator-line"
      style="--vertical-indicator-color: var(--grey-40); left: 266.6666666666667px;"
    />
    <div
      class="timelineVerticalIndicatorsLine"
      data-marker-index="14"
      data-testid="vertical-indicator-line"
      style="--vertical-indicator-color: var(--magenta-60); left: 333.33333333333337px;"
    />
  </div>
</div>
`;
